xclim/sdba/properties.py

```diff
@@ -947,6 +1004,147 @@ def _spatial_correlogram(da: xr.DataArray, *, dims=None, bins=100, group="time")
 )
 
 
+def _decorrelation_length(
+    da: xr.DataArray, *, radius=300, thresh=0.50, dims=None, bins=100, group="time"
+):
+    """Decorrelation length.
+
+    Distance from a grid cell where the correlation with its neighbours goes below the threshold.
+    A correlogram is calculated for each grid cell following the method from ``xclim.sdba.properties.spatial_correlogram``.
+    Then, we find the first bin closest to the correlation threshold.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+      Data.
+    radius: int
+        Radius (in km) defining the region where correlations will be calculate between a point and its neighbours.
+    thresh: float
+      Threshold correlation defining decorrelation.
+       The decorrelation length is defined as the center of the distance bin that has a correlation closest to this threshold.
+    dims: sequence of strings
+      Name of the spatial dimensions. Once these are stacked, the longitude and latitude coordinates must be 1D.
+    bins:
+      Same as argument `bins` from :py:meth:`scipy.stats.binned_statistic`.
+      If given as a scalar, the equal-width bin limits from 0 to radius are generated here
+      (instead of letting scipy do it) to improve performance.
+    group: str
+      Useless for now.
+
+    Returns
+    -------
+    xr.DataArray, [dimensionless]
+      Decorrelation length.
+
+    Notes
+    -----
+    Calculating this property requires a lot of memory. It will not work with large datasets.
+    """
+    # save attrs
+    coords_attrs = {}
+    for c in da.coords:
+        coords_attrs[c] = da[c].attrs
+
+    if dims is None:
+        dims = [d for d in da.dims if d != "time"]
+
+    corr = _pairwise_spearman(da, dims)
+
+    dists, mn, mx = _pairwise_haversine_and_bins(
+        corr.cf["longitude"].values, corr.cf["latitude"].values
+    )
+
+    # fill bottom triangle
+    np.nan_to_num(dists, 0)
+    # TODO: do it in 2 shots instead?
+    dists = dists + dists.T  # - np.diag(np.diag(dists))
```

Comment:
I think this could be more efficient if done in `sdba.nbutils._pairwise_haverinse_and_bins`?

```python3
@njit
def _pairwise_haversine_and_bins2(lond, latd, transpose=False):
    """Inter-site distances with the haversine approximation."""
    N = lond.shape[0]
    lon = np.deg2rad(lond)
    lat = np.deg2rad(latd)
    dists = np.full((N, N), np.nan)
    for i in range(N - 1):
        for j in range(i + 1, N):
            dlon = lon[j] - lon[i]
            dists[i, j] = 6367 * np.arctan2(
                np.sqrt(
                    (np.cos(lat[j]) * np.sin(dlon)) ** 2
                    + (
                        np.cos(lat[i]) * np.sin(lat[j])
                        - np.sin(lat[i]) * np.cos(lat[j]) * np.cos(dlon)
                    )
                    ** 2
                ),
                np.sin(lat[i]) * np.sin(lat[j])
                + np.cos(lat[i]) * np.cos(lat[j]) * np.cos(dlon),
            )
            if transpose:
                dists[j, i] = dists[i, j]
    mn = np.nanmin(dists)
    mx = np.nanmax(dists)
    if transpose:
        np.fill_diagonal(dists, 0)
    return dists, mn, mx
```

With 1000 points, I get 50 ms with this code, no matter if `transpose` is True or False, which is the same as with the previous version. And 65 ms with your version. I thought the difference would be larger, but here is my code now that I did the test...